### PR TITLE
Add test case with an advance() method to spool forward time

### PR DESCRIPTION
**Pull request by foolswood**
Useful for testing stuff with timers.

Same sort of idea as Clock in twisted.trial if you're familiar with that.

### CODE REVIEW

File: asynctest/case.py [hunk 1]
```diff
@@ -340,6 +340,26 @@ class FunctionTestCase(TestCase, unittest.FunctionTestCase):
     """
 
 
+class ClockedTestCase(TestCase):
+    def setUp(self):
+        super().setUp()
+        self.loop.time = lambda: self.time
```
Comment (Martiusweb) [inline asynctest/case.py]: I'd ``functools.wraps()`` the original method.

File: asynctest/case.py [hunk 2]
```diff
@@ -340,6 +340,26 @@ class FunctionTestCase(TestCase, unittest.FunctionTestCase):
...
+        assert seconds >= 0
+        self._drain_loop()
+        self.time += seconds
+        self._drain_loop()
+
+    def _drain_loop(self):
+        while self.loop._ready or any(map(
+                lambda handle: handle._when <= self.time,
+                self.loop._scheduled)):
+            self.loop._run_once()
```
Comment (Martiusweb) [inline asynctest/case.py]: The loop is not re-entrant, so ``advance()`` can't be called on top of the loop (from a coroutine test, for example), transforming ``_drain_loop()`` and ``advance()`` into coroutine functions and ``await  asyncio.sleep(0)`` until there are no callbacks left to run is probably a better solution (like with ``asyncio.helpers.exhaust_loop``).

On the other hand, would it be a bad idea to run ``advance()`` asynchronously during the test?

File: asynctest/case.py [hunk 3]
```diff
@@ -340,6 +340,26 @@ class FunctionTestCase(TestCase, unittest.FunctionTestCase):
...
 
 
+class ClockedTestCase(TestCase):
+    def setUp(self):
+        super().setUp()
+        self.loop.time = lambda: self.time
+        self.time = 0
+
+    def advance(self, seconds):
+        assert seconds >= 0
```
Comment (Martiusweb) [inline asynctest/case.py]: Hm, asserting in a test would mark the test as failing instead of "got errors", I'd raise ``ValueError``.

**Comment by Martiusweb** [comment]
Hi again. It's a great idea.

I added several comments to your patch that we should discuss.

It seems that twisted add ``advance()`` to is base ``TestCase`` class, we may want to do the same, instead of subclassing.

Thanks!

**Comment by foolswood** [comment]
The twisted base TestCase doesn't have an advance() method (see https://github.com/twisted/twisted/blob/trunk/twisted/trial/_asynctest.py) what twisted does is produce a different event loop (the clock) in which you control time.

With this either the wall clock time controls the time, or the developer does. The test author has to take into account the managing of time when using this, putting it on the base forces everyone to do it.

You could have both in control of the time at once in theory, but I think that'd probably result in racy tests.

File: asynctest/case.py (hunk 2 above)
Comment (foolswood) [inline asynctest/case.py]: I'm not sure, what I've tended to find is what you normally end up writing is something like:
- set up conditions
- assertions
- simulate wait
- assertions

So I guess it depends on how you're writing the tests.
If you're using run_until_complete then it'll be easier as is, otherwise the coroutine will be better.
I am happy to go with whichever is more idiomatic for the lib.

**Comment by Martiusweb** [comment]
Ok, a `ClockedTestCase` makes sense then.

When I read tests for asyncio libraries, usually authors use coroutine functions as their test cases (they use `@pytest.mark.asyncio`, https://github.com/pytest-dev/pytest-asyncio) or write their own decorator which runs the function on the loop. Can you show me an example of how you write your test functions with asynctest?

I think it's better to make `advance` a coroutine function, as one can still write `loop.run_until_complete(self.advance(1))` if the test is a function, while the converse is not possible.

Can you also add a test which asserts that `ValueError` is raised when the argument is negative?

**Comment by foolswood** [comment]
We came to the asynctest party late, so we have synchronous tests with run_until_complete scattered through. Just wasn't sure what the intention was. Will get on and make the suggested changes.

**Comment by Martiusweb** [comment]
Can you add those two tests (or assertions in test_advance):
* one ensuring that a callback scheduled for after 2 seconds is not yet executed after `advance(1)` and is executed if we then call `advance(3)` (ensuring the callback doesn't run too soon and that we don't need an exact value to call the callback).
* check the Handle object returned by loop.call_later() and ensure it is not done before the call to advance() but is after.

We also need documentation for `ClockedTestCase` and `advance()`.

I think that once it's done, we'll be ready to merge the PR!

**Comment by Martiusweb** [comment]
Ok, everything works!

I also checked that it works with `loop.run_until_complete(self.advance(1))` and when combined with `asyncio.sleep()`.

I merge the PR, thank you, this is the first feature introduced in asyncio from a contribution!

**Comment by foolswood** [comment]
Cool, thanks for being so responsive!
